contracts/staking/ClusterRewards.sol

```diff
@@ -235,161 +231,217 @@ contract ClusterRewards is
-            bytes32 _networkId, 
-            uint256 _fromEpoch, 
-            uint256 _noOfEpochs,
-            uint16[][] memory _tickets
-        ) = _parseTicketInfo(_ticketInfo);
+        (bytes32 _networkId, uint256 _fromEpoch, uint256 _noOfEpochs, uint16[][] memory _tickets) = _parseTicketInfo(_ticketInfo);
 
         ReceiverStaking _receiverStaking = receiverStaking;
         require(_fromEpoch + _noOfEpochs <= _receiverStaking.getCurrentEpoch(), "CRW:ITC-Epochs not completed");
 
         uint256[] memory _stakes = _receiverStaking.totalSupplyAtRanged(_fromEpoch, _noOfEpochs);
         (uint256[] memory _balances, address _receiver) = _receiverStaking.balanceOfSignerAtRanged(msg.sender, _fromEpoch, _noOfEpochs);
         address[][] memory _selectedClusters = clusterSelectors[_networkId].getClustersRanged(_fromEpoch, _noOfEpochs);
-        uint256 _epochLength = _receiverStaking.EPOCH_LENGTH();
+
+        uint256 _receiverExtraRewardsRemaining = receiverBalance[_receiver];
+        uint256 _receiverExtraRewardsPerEpoch = receiverRewardPerEpoch[_receiver];
         uint256 _totalNetworkRewardsPerEpoch;
 
         unchecked {
-            for(uint256 i=0; i < _noOfEpochs; ++i) {
-                _totalNetworkRewardsPerEpoch = _getRewardForEpoch(_fromEpoch, _networkId, _epochLength);
-                _processReceiverTickets(_receiver, _fromEpoch, _selectedClusters[i], _tickets[i], _totalNetworkRewardsPerEpoch, _stakes[i], _balances[i]);
-                emit TicketsIssued(_networkId, _fromEpoch, msg.sender);
+            for (uint256 i = 0; i < _noOfEpochs; ++i) {
+                // _totalNetworkRewardsPerEpoch = inflation rewards + receiver rewards
+                _totalNetworkRewardsPerEpoch = getRewardForEpoch(_fromEpoch, _networkId);
+                uint256 _rewardShare = _getRewardShare(_totalNetworkRewardsPerEpoch, _stakes[i], _balances[i]) +
+                    _extraReceiverRewardThisEpoch(_receiverExtraRewardsRemaining, _receiverExtraRewardsPerEpoch);
+
+                _processReceiverTickets(_receiver, _fromEpoch, _selectedClusters[i], _tickets[i], _rewardShare);
+                _emitTicketsIssued(_networkId, _fromEpoch, msg.sender);
+                _receiverExtraRewardsRemaining -= _extraReceiverRewardThisEpoch(
+                    _receiverExtraRewardsRemaining,
+                    _receiverExtraRewardsPerEpoch
+                );
                 ++_fromEpoch;
             }
         }
+
+        _setReceiverBalance(_receiver, _receiverExtraRewardsRemaining);
     }
 
-    function _parseTicketInfo(bytes memory ticketInfo) internal view returns(
-        bytes32 networkId,
-        uint256 fromEpoch,
-        uint256 noOfEpochs,
-        uint16[][] memory tickets
-    ) {
+    function _setReceiverBalance(address _receiver, uint256 _receiverExtraRewardsRemaining) internal {
+        if (_receiverExtraRewardsRemaining != 0) {
+            receiverBalance[_receiver] = _receiverExtraRewardsRemaining;
+        }
```

Comment:
if _receiverExtraRewardsRemaining == 0, then receiverBalance should be updated as well right ?

Reply:
ha ,
was a external function before..
Changed